src/image/VisitInfo.cc

```diff
@@ -472,11 +472,31 @@ VisitInfo::VisitInfo(daf::base::PropertySet const& metadata)
     }
 }
 
+/**
+ * Test whether two numbers are exactly equal or both NaN.
+ *
+ * This function is exactly equivalent to `operator==(double, double)`, except
+ * that NaNs compare equal (NaNs of different values also compare equal).
+ */
+bool _eqOrNan(double lhs, double rhs) noexcept { return (std::isnan(lhs) && std::isnan(rhs)) || lhs == rhs; }
+
+/**
+ * Test whether two SpherePoints are exactly equal or invalid.
+ *
+ * This function is needed because SpherePoint::operation== is specifically
+ * designed to be ill-behaved when NaN values are involved.
+ */
+bool _eqOrNan(lsst::geom::SpherePoint const& lhs, lsst::geom::SpherePoint const& rhs) noexcept {
+    return (!lhs.isFinite() && !rhs.isFinite()) || lhs == rhs;
+}
+
 bool VisitInfo::operator==(VisitInfo const& other) const {
-    return _exposureId == other.getExposureId() && _exposureTime == other.getExposureTime() &&
-           _darkTime == other.getDarkTime() && _date == other.getDate() && _ut1 == other.getUt1() &&
-           _era == other.getEra() && _boresightRaDec == other.getBoresightRaDec() &&
-           _boresightAzAlt == other.getBoresightAzAlt() && _boresightAirmass == other.getBoresightAirmass() &&
+    return _exposureId == other.getExposureId() && _eqOrNan(_exposureTime, other.getExposureTime()) &&
```

Comment:
This mostly makes me wonder why `_id == other.getId()` is not sufficient :) Not worth digging into that here.